src/sonic-config-engine/minigraph.py

```diff
@@ -98,7 +99,92 @@ def parse_device(device):
             deployment_id = node.text
     return (lo_prefix, lo_prefix_v6, mgmt_prefix, name, hwsku, d_type, deployment_id)
 
-def parse_png(png, hname):
+def calculate_lcm_for_ecmp (nhdevices_bank_map, nhip_bank_map):
+    banks_enumerated = {}
+    lcm_array = []
+    for value in nhdevices_bank_map.values():
+        for key in nhip_bank_map.keys():
+            if nhip_bank_map[key] == value:
+                if value not in banks_enumerated:
+                    banks_enumerated[value] = 1
+                else:
+                    banks_enumerated[value] = banks_enumerated[value] + 1
+    for bank_enumeration in banks_enumerated.values():
+        lcm_list = range(1, bank_enumeration+1)
+        lcm_comp = lcm_list[0]
+        for i in lcm_list[1:]:
+            lcm_comp = lcm_comp * i / calculate_gcd(lcm_comp, i)
+        lcm_array.append(lcm_comp)
+
+    LCM = sum(lcm_array)
+    return int(LCM)
+
+def calculate_gcd(x, y):
+    while y != 0:
+        (x, y) = (y, x % y)
+    return int(x)
+
+def formulate_ecmp_entry(dpg_ecmp_content, port_device_map):
+    png_ecmp_content = {}
+    FG_NHG_MEMBER = [{}, {}]
+    FG_NHG_PREFIX = [{}, {}]
+    FG_NHG = [{}, {}]
+    FG_NEIGH = [{}, {}]
+    COMP_FG_NHG_MEMBER = {}
+    COMP_FG_NHG_PREFIX = {}
+    COMP_FG_NHG = {}
+    COMP_FG_NEIGH = {}
+    neigh_key_ipv4 = []
+    neigh_key_ipv6 = []
+    ipv4_tag = "fgnhg_v4"
+    ipv6_tag = "fgnhg_v6"
+    port_nhipv4_map = dpg_ecmp_content['port_nhipv4_map']
+    port_nhipv6_map = dpg_ecmp_content['port_nhipv6_map']
+    nhgaddr = dpg_ecmp_content['nhgaddr']
+    nhg_int = dpg_ecmp_content['nhgint']
+    nhipv4_device_map = {port_nhipv4_map[x]: port_device_map[x] for x in port_device_map
+                         if x in port_nhipv4_map}
+    nhipv6_device_map = {port_nhipv6_map[x]: port_device_map[x] for x in port_device_map
+                         if x in port_nhipv6_map}
+    nhipv4_devices = sorted(list(set(nhipv4_device_map.values())))
+    nhipv6_devices = sorted(list(set(nhipv6_device_map.values())))
+
+    nhdevices_ipv4_bank_map = {device: bank for bank, device in enumerate(nhipv4_devices)}
+    nhdevices_ipv6_bank_map = {device: bank for bank, device in enumerate(nhipv6_devices)}
+
+    nhipv4_bank_map = {ip: nhdevices_ipv4_bank_map[device] for ip, device in nhipv4_device_map.items()}
+    nhipv6_bank_map = {ip: nhdevices_ipv6_bank_map[device] for ip, device in nhipv6_device_map.items()}
+
+    ipv4_LCM = calculate_lcm_for_ecmp(nhdevices_ipv4_bank_map, nhipv4_bank_map)
+    ipv6_LCM = calculate_lcm_for_ecmp(nhdevices_ipv6_bank_map, nhipv6_bank_map)
+
+    FG_NHG_MEMBER[0] = {ip: {"FG_NHG": ipv4_tag, "bank": bank} for ip, bank in nhipv4_bank_map.items()}
+    FG_NHG_PREFIX[0] = {nhgaddr[0]: {"FG_NHG": ipv4_tag}}
+    FG_NHG[0] = {ipv4_tag: {"bucket_size": ipv4_LCM}}
+    for ip in nhipv4_bank_map:
+        neigh_key_ipv4.append(str(nhg_int + "|" + ip))
```

Comment:
can we make this logic more generic ? instead of doing v4 and v6 separately can we look on ip type and do same thing ?